### ISSUE TITLE
`cupy.fuse` unable to promote Python scalars in some cases (NEP 50?)

### ISSUE THREAD

**Issue by jakirkham** [cupy/cupy#8536]
### Description

A colleague identified an issue where Python scalars are sometimes not promoted to the appropriate device type ( https://github.com/rapidsai/cucim/issues/742#issuecomment-2299877295 ). Have included the code he used below

Though have also reproduced the same issue on Google Colab

### To Reproduce

Code:

```py
import cupy as cp

delta_phi = cp.full((10, 10), 9.999999e-7, dtype=cp.float32)
dt = cp.asarray(0.5, dtype=cp.float32)
mu = cp.asarray(0.1, dtype=cp.float32)

@cp.fuse()
def fused_kernel(dt, mu, delta_phi):
    return 1 + mu * dt * delta_phi

fused_kernel(dt, mu, delta_phi)
```

Error message:

```python
Traceback (most recent call last):
  File "cupy/_core/fusion.pyx", line 891, in cupy._core.fusion.Fusion.__call__
  File "cupy/_core/fusion.pyx", line 732, in cupy._core.fusion._FusionHistory.get_fusion
  File "<string>", line 10, in fused_kernel
  File "cupy/_core/fusion.pyx", line 198, in cupy._core.fusion._FusionVarScalar.__radd__
  File "cupy/_core/_kernel.pyx", line 1241, in cupy._core._kernel.ufunc.__call__
  File "cupy/_core/_fusion_thread_local.pyx", line 35, in cupy._core._fusion_thread_local.call_ufunc
  File "cupy/_core/fusion.pyx", line 949, in cupy._core.fusion._call_ufunc
  File "cupy/_core/fusion.pyx", line 605, in cupy._core.fusion._FusionHistory.call_ufunc
  File "cupy/_core/fusion.pyx", line 554, in cupy._core.fusion._FusionHistory.call_ufunc.can_cast1
TypeError: can_cast() does not support Python ints, floats, and complex because the result used to depend on the value.
This change was part of adopting NEP 50, we may explicitly allow them again in the future.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 12, in <module>
  File "cupy/_core/fusion.pyx", line 897, in cupy._core.fusion.Fusion.__call__
  File "cupy/_core/new_fusion.pyx", line 162, in cupy._core.new_fusion.Fusion.__call__
  File "cupy/_core/_fusion_kernel.pyx", line 353, in cupy._core._fusion_kernel.FusedKernel.execute
  File "cupy/_util.pyx", line 64, in cupy._util.memoize.decorator.ret
  File "cupy/_core/_fusion_kernel.pyx", line 49, in cupy._core._fusion_kernel._cuda_compile
  File "cupy/_core/core.pyx", line 2271, in cupy._core.core.compile_with_cache
  File "/content/cupy_env/lib/python3.10/site-packages/cupy/cuda/compiler.py", line 495, in _compile_module_with_cache
    return _compile_with_cache_cuda(
  File "/content/cupy_env/lib/python3.10/site-packages/cupy/cuda/compiler.py", line 580, in _compile_with_cache_cuda
    ls.add_ptr_data(ptx, 'cupy.ptx')
  File "cupy/cuda/function.pyx", line 294, in cupy.cuda.function.LinkState.add_ptr_data
  File "cupy/cuda/function.pyx", line 295, in cupy.cuda.function.LinkState.add_ptr_data
  File "cupy_backends/cuda/api/driver.pyx", line 179, in cupy_backends.cuda.api.driver.linkAddData
  File "cupy_backends/cuda/api/driver.pyx", line 63, in cupy_backends.cuda.api.driver.check_status
cupy_backends.cuda.api.driver.CUDADriverError: CUDA_ERROR_NO_BINARY_FOR_GPU: no kernel image is available for execution on the device
```

### Installation

Built from GitHub source

### Environment

Used a build of CuPy from the `v13` branch in conda-forge

```
OS                           : Linux-6.1.85+-x86_64-with-glibc2.35
Python Version               : 3.10.14
CuPy Version                 : 13.2.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 2.1.0
SciPy Version                : None
Cython Build Version         : 0.29.37
Cython Runtime Version       : None
CUDA Root                    : /usr/local/cuda
nvcc PATH                    : /usr/local/cuda/bin/nvcc
CUDA Build Version           : 12050
CUDA Driver Version          : 12020
CUDA Runtime Version         : 12050 (linked to CuPy) / 12020 (locally installed)
CUDA Extra Include Dirs      : ['/content/cupy_env/targets/x86_64-linux/include', '/content/cupy_env/include']
cuBLAS Version               : (available)
cuFFT Version                : 11203
cuRAND Version               : 10306
cuSOLVER Version             : (11, 6, 3)
cuSPARSE Version             : (available)
NVRTC Version                : (12, 5)
Thrust Version               : 200400
CUB Build Version            : 200600
Jitify Build Version         : 1a0ca0e
cuDNN Build Version          : 8907
cuDNN Version                : 8906
NCCL Build Version           : 22203
NCCL Runtime Version         : 21903
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : Tesla T4
Device 0 Compute Capability  : 75
Device 0 PCI Bus ID          : 0000:00:04.0
```

### Additional Information

We are not sure whether this is expected behavior

**Comment by jakirkham** [comment]
FWIW we did find a workaround. Basically cast `1` to match the same type as the variables, which we are happy to do

So at this point just trying to understand if this should have changed. Or if there is unexpected

**Comment by seberg** [comment]
I think the surprising thing is that we didn't see it before.  But there is the old/new fusion path, so it might be related to that.
(I had seen similar things before locally, that errors only showed up because the new path also failed, IIRC.)

I suspect that in v14 (the NEP 50 PR), this is already effectively fixed, but adding a work-around for NumPy 2 + v13.3 is likely easy enough that I'll try to have a look.

**Comment by jcrist** [comment]
We just ran into this issue in `cuml` - here's a self contained reproducer:

```python
import cupy as cp


def call(func, *args):
    try:
        func(*args)
    except Exception:
        print(f"{func.__name__} errored!")
    else:
        print(f"{func.__name__} succeeded!")


@cp.fuse()
def original(x, h):
    """The code we'd like to write - this fails"""
    return 1.0 - (x * x) / (h * h)


@cp.fuse()
def scalar_passed_in(x, h, one):
    """Moving the scalar to be passed in as an arg succeeds"""
    return one - (x * x) / (h * h)


@cp.fuse()
def h_squared_passed_in(x, h, h_squared):
    """Moving `h * h` (a scalar) to be passed in as an arg succeeds"""
    return 1.0 - (x * x) / h_squared


X = cp.random.normal(size=(10, 20)).astype("float32")
h = 5.0

print(f"Using cupy version {cp.__version__}")
call(original, X, h)
call(scalar_passed_in, X, h, 1.0)
call(h_squared_passed_in, X, h, h * h)
```

And the output:

```
Using cupy version 13.3.0
original errored!
scalar_passed_in succeeded!
h_squared_passed_in succeeded!
```

It's not clear to me when this bug crops up - sometimes scalars in the elementwise operations are fine, and sometimes they lead to the error seen above. For now we have a workaround (went with `h_squared_passed_in` above).

**Comment by seberg** [comment]
Sorry, I think I never looked at it closer because I, terribly so, thought that cupy 14.0 was coming and already had the fix.

Maybe as a quick note: the other work-around is to use `np.float32` as the scalar value (assuming it isn't a value that would lose precision).  That way you get strong promotion.  Since I presume the input is only ever float32 or float64, that should be correct.

I'll have a look, probably my comment above already contains the minimal solution.

---

> It's not clear to me when this bug crops up

There are two fused code paths with a `try/except` which may make it hard to follow sometimes (and possible is the reason for this being missed to begin with).

**Comment by jakirkham** [comment]
AIUI this was fixed in `13.4.1` with PR: https://github.com/cupy/cupy/pull/9012

The code path is different in `14.x`. So doesn't present the same issue. A test was added for `14.x` with PR: https://github.com/cupy/cupy/pull/9013

Closing as resolved. For any new issues, please file a new report using `13.4.1`

Thanks all! 🙏
